packages/react-router/__tests__/router/navigation-blocking-test.ts

```diff
@@ -442,21 +454,58 @@ describe("navigation blocking", () => {
         router.getBlocker("KEY", fn);
         await router.navigate(-1);
         router.getBlocker("KEY", fn).proceed?.();
-        await sleep(LOADER_LATENCY_MS);
-        expect(router.getBlocker("KEY", fn)).toEqual({
-          state: "unblocked",
-          proceed: undefined,
-          reset: undefined,
-          location: undefined,
-        });
+        await waitFor(() =>
+          expect(router.getBlocker("KEY", fn)).toEqual({
+            state: "unblocked",
+            proceed: undefined,
+            reset: undefined,
+            location: undefined,
+          })
+        );
       });
 
       it("navigates after proceeding navigation completes", async () => {
         router.getBlocker("KEY", fn);
         await router.navigate(-1);
         router.getBlocker("KEY", fn).proceed?.();
-        await sleep(LOADER_LATENCY_MS);
-        expect(router.state.location.pathname).toBe("/about");
+        await waitFor(() =>
+          expect(router.state.location.pathname).toBe("/about")
+        );
+      });
+    });
+
+    describe("proceeds from blocked state using browser history", () => {
+      let fn = () => true;
+
+      // we want to navigate so that `/about` is the previous entry in the
+      // stack here since it has a loader that won't resolve immediately
+      beforeEach(async () => {
+        const history = createBrowserHistory();
+
+        router = createRouter({
+          history,
+          routes,
+        });
+
+        router.initialize();
+
+        await router.navigate("/");
+        await router.navigate("/about");
+        await router.navigate("/contact");
+      });
+
+      it("proceeds after quick block of back navigation", async () => {
+        router.getBlocker("KEY", fn);
+
+        await router.navigate(-1); // This does not really wait for the navigation to happen
+        await waitFor(
+          () => expect(router.getBlocker("KEY", fn).state).toBe("blocked"),
+          { interval: 1 }
+        ); // This awaits the navigation
+        router.getBlocker("KEY", fn).proceed!();
+        await waitFor(() =>
+          expect(router.state.location.pathname).toBe("/about")
```

Comment:
I am not sure why it behaves differently here. My assumption was that maybe JSDOM (or whoever emulates browser navigation here) behaves differently in the case where navigation is in progress and you navigate again. I think it still covers the issue though, which is that if you navigate during navigation, you will end up somewhere else than you would expect.

If you have ideas on how to improve the test, I am all ears

Reply:
Ah ok yeah this is probably a JSDOM issue.  I'm going to test this through an integration test in a real browser in https://github.com/remix-run/remix/pull/9914 instead of trying to hack JSDOM into behaving correctly